### PR TITLE
Add checker and authorizer.

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

File: macaroonbakery/checker.py [hunk 1]
```diff
@@ -36,4 +49,388 @@ def __hash__(self):
...
+        self._macaroons = macaroons
+        self._init_errors = []
+        self._executed = False
+        self._identity = None
+        self._identity_caveats = []
+        self.parent = parent
+        self._conditions = None
+
+    def _init(self, ctx):
+        if not self._executed:
```
Comment (rogpeppe) [inline macaroonbakery/checker.py]: Are we sure this won't be executed in a multi-threaded way?

File: macaroonbakery/checker.py [hunk 2]
```diff
@@ -36,4 +49,388 @@ def __hash__(self):
...
+                msg='authentication required',
+                ops=[LOGIN_OP],
+                cavs=self._identity_caveats)
+        if caveats is None or len(caveats) == 0:
+            all_errors = []
+            all_errors.extend(self._init_errors)
+            all_errors.extend(errors)
+            err = ''
+            if len(all_errors) > 0:
+                err = ' '.join(all_errors)
```
Comment (rogpeppe) [inline macaroonbakery/checker.py]: Joining the errors with spaces doesn't seem quite right to me. I haven't yet figured out a decent way to format the errors when there's more than one, so perhaps just choose the first one like the Go version does?

File: macaroonbakery/checker.py [hunk 3]
```diff
@@ -36,4 +49,388 @@ def __hash__(self):
...
+        returned identity_client. It may be None after a successful
+        authorization if LOGIN_OP access was not required.
+        :param macaroons: holds all the macaroons that were used for the
+        authorization. Macaroons that were invalid or unnecessary are
+        not included.
+        '''
+        return super(AuthInfo, cls).__new__(cls, identity, macaroons)
+
+
+class _CaveatSquasher(object):
```
Comment (rogpeppe) [inline macaroonbakery/checker.py]: Please include a doc comment on this type.

File: macaroonbakery/checker.py [hunk 4]
```diff
@@ -36,4 +49,388 @@ def __hash__(self):
...
+        return True
+
+    def final(self):
+        if self._expiry is not None:
+            self._conds.append(time_before_caveat(self._expiry).condition)
+        if len(self._conds) == 0:
+            return []
+        # Make deterministic and eliminate duplicates.
+        self._conds = sorted(set(self._conds))
+        prev = self._conds[0]
```
Comment (rogpeppe) [inline macaroonbakery/checker.py]: I think that all this code is redundant because you've used set() above.

File: macaroonbakery/codec.py [hunk 5]
```diff
@@ -8,9 +8,9 @@
 from nacl.encoding import Base64Encoder
 from nacl.public import Box, PublicKey
 
-import macaroonbakery
-from macaroonbakery import bakery
-from macaroonbakery.checkers.namespace import deserialize_namespace
+from macaroonbakery import BAKERY_V1, BAKERY_V2, BAKERY_V3
+from macaroonbakery.third_party import legacy_namespace, ThirdPartyCaveatInfo
+from macaroonbakery.checkers import deserialize_namespace
```
Comment (rogpeppe) [inline macaroonbakery/codec.py]: Maybe just import macaroonbakery.checkers ?

File: macaroonbakery/discharge.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
+import abc
+from collections import namedtuple
+
+
+from macaroonbakery.checkers import (
```
Comment (rogpeppe) [inline macaroonbakery/discharge.py]: import checkers ?

File: macaroonbakery/discharge.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        # about to throw it away.
+        discharge_m = dm.macaroon()
+        m = primary.prepare_for_request(discharge_m)
+        discharges.append(m)
+        add_caveats(dm)
+    return discharges
+
+
+class ThirdPartyCaveatChecker(object):
+    ''' Holds a function that checks third party caveats for validity.
```
Comment (rogpeppe) [inline macaroonbakery/discharge.py]: Defines an abstract class that's used to check third party caveats.

then define the behaviour of the check_third_party_caveat method inside that method?
?

File: macaroonbakery/discharge.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    If the caveat is valid, it returns a nil error and optionally a slice of
+    extra caveats that will be added to the discharge macaroon.
+    The caveatId parameter holds the still-encoded id of the caveat.
+    If the caveat kind was not recognised, the checker should return an
+    error with a ErrCaveatNotRecognized cause.
+    '''
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def check_third_party_caveat(self, ctx, info):
```
Comment (rogpeppe) [inline macaroonbakery/discharge.py]: This should have a doc comment that documents its argument and return types.

File: macaroonbakery/discharge.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+class _LocalDischargeChecker(ThirdPartyCaveatChecker):
+    def check_third_party_caveat(self, ctx, info):
+        if info.condition.decode('utf-8') != 'true':
+            raise ErrCaveatNotRecognized()
+        return None
+
+
+def discharge(ctx, id, caveat, key, checker, locator):
+    ''' Creates a macaroon to discharges a third party caveat.
```
Comment (rogpeppe) [inline macaroonbakery/discharge.py]: s/discharges/discharge/

File: macaroonbakery/discharge.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+        declared[parts[0]] = True
+    # Add empty declarations for everything mentioned in need-declared
+    # that was not actually declared.
+    for d in need_declared:
+        if not declared[d]:
+            caveats.append(declared_caveat(d, ''))
+    return caveats
+
+
+class _EmptyLocator(object):
```
Comment (rogpeppe) [inline macaroonbakery/discharge.py]: shouldn't this inherit from the ThirdPartyLocator abstract base class?

File: macaroonbakery/error.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
+class DischargeRequiredError(Exception):
+    ''' Raised when authorization has failed and a discharged macaroon might
```
Comment (rogpeppe) [inline macaroonbakery/error.py]: Raised by Checker ?

File: macaroonbakery/error.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        self._cavs = cavs
+
+    def ops(self):
+        return self._ops
+
+    def cavs(self):
+        return self._cavs
+
+
+class PermissionDenied(Exception):
```
Comment (rogpeppe) [inline macaroonbakery/error.py]: doc comments for these errors please.

File: macaroonbakery/identity.py [hunk 13]
```diff
@@ -72,3 +72,48 @@ def allow(self, acls):
             if self._identity == acl:
                 return True
         return False
+
+
+class IdentityClient(object):
```
Comment (rogpeppe) [inline macaroonbakery/identity.py]: doc comment please.

File: macaroonbakery/identity.py [hunk 14]
```diff
@@ -72,3 +72,48 @@ def allow(self, acls):
...
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def identity_from_context(self, ctx):
+        ''' Returns the identity based on information in the context.
```
Comment (rogpeppe) [inline macaroonbakery/identity.py]: Please document the argument and return types.

File: macaroonbakery/identity.py [hunk 15]
```diff
@@ -72,3 +72,48 @@ def allow(self, acls):
...
+        It should only return an error if it cannot check the identity
+        (for example because of a database access error) - it's
+        OK to return all zero values when there's
+        no identity found and no third party to address caveats to.
+        '''
+        raise NotImplementedError('identity_from_context method must be '
+                                  'defined in subclass')
+
+    @abc.abstractmethod
+    def declared_identity(self, ctx, declared):
```
Comment (rogpeppe) [inline macaroonbakery/identity.py]: Please document arg and return types.

File: macaroonbakery/identity.py [hunk 16]
```diff
@@ -72,3 +72,48 @@ def allow(self, acls):
...
+        '''Parses the identity declaration from the given declared attributes.
+
+        TODO take the set of first party caveat conditions instead?
+        '''
+        raise NotImplementedError('declared_identity method must be '
+                                  'defined in subclass')
+
+
+class IdentityError(Exception):
```
Comment (rogpeppe) [inline macaroonbakery/identity.py]: doc comment

File: macaroonbakery/keys.py [hunk 17]
```diff
@@ -0,0 +1,7 @@
+from nacl.public import PrivateKey
+
+
+def generate_key():
+    '''GenerateKey generates a new key pair.
```
Comment (rogpeppe) [inline macaroonbakery/keys.py]: Please document the return type.

File: macaroonbakery/macaroon.py [hunk 18]
```diff
@@ -8,27 +8,22 @@
...
-    '''
-    ns = checkers.Namespace(None)
-    ns.register(checkers.STD_NAMESPACE, '')
-    return ns
+log = logging.getLogger(__name__)
 
 
-class Macaroon:
+class Macaroon(object):
     '''Represent an undischarged macaroon along its first
```
Comment (rogpeppe) [inline macaroonbakery/macaroon.py]: s/its/with its/

File: macaroonbakery/macaroon.py [hunk 19]
```diff
@@ -157,7 +167,38 @@ def _new_caveat_id(self, base):
         @param base string
         @return string
         '''
-        raise NotImplementedError
+        id = bytearray()
+        if len(base) > 0:
+            id.append(base)
+        else:
+            # Add a version byte to the caveat id.Technically
```
Comment (rogpeppe) [inline macaroonbakery/macaroon.py]: s/Tech/ Tech/

File: macaroonbakery/macaroon.py [hunk 20]
```diff
@@ -157,7 +167,38 @@ def _new_caveat_id(self, base):
...
+            # this is unnecessary as the caveat-decoding logic
+            # that looks at versions should never see this id,
+            # but if the caveat payload isn't provided with the
+            # payload, having this version gives a strong indication
+            # that the payload has been omitted so we can produce
+            # a better error for the user.
+            id.append(BAKERY_V3)
+        id.append(0)
```
Comment (rogpeppe) [inline macaroonbakery/macaroon.py]: What's this for?

File: macaroonbakery/macaroon.py [hunk 21]
```diff
@@ -157,7 +167,38 @@ def _new_caveat_id(self, base):
         @param base string
         @return string
```
Comment (rogpeppe) [inline macaroonbakery/macaroon.py]: I think this should be taking and returning bytes not strings.

File: macaroonbakery/macaroon.py [hunk 22]
```diff
@@ -157,7 +167,38 @@ def _new_caveat_id(self, base):
...
+            # is using similar conventions - in the worst case they might
+            # end up with a duplicate third party caveat id and thus create
+            # a macaroon that cannot be discharged.
+            temp[:] = id
+            codec.encode_uvarint(i, temp)
+            for cav in caveats:
+                if cav.verification_id is not None and cav.caveat_id == temp:
+                    i += 1
+                    break
+            return temp.decode('utf-8')
```
Comment (rogpeppe) [inline macaroonbakery/macaroon.py]: This doesn't seem right to me, as not all uvarints can be encoded as utf-8.
Shouldn't this be bytes(temp) ?

File: macaroonbakery/macaroon.py [hunk 23]
```diff
@@ -157,7 +167,38 @@ def _new_caveat_id(self, base):
...
+            # a better error for the user.
+            id.append(BAKERY_V3)
+        id.append(0)
+        # Iterate through integers looking for one that isn't already used,
+        # starting from n so that if everyone is using this same algorithm,
+        # we'll only perform one iteration.
+        i = len(self._caveat_data)
+        caveats = self._macaroon.caveats
+        while True:
+            temp = bytearray()
```
Comment (rogpeppe) [inline macaroonbakery/macaroon.py]: temp = id[:]
?

File: macaroonbakery/checkers/declared.py [hunk 24]
```diff
@@ -3,7 +3,7 @@
 
 from macaroonbakery.checkers.namespace import Namespace
 from macaroonbakery.checkers.caveat import parse_caveat
-from macaroonbakery.checkers import COND_DECLARED, STD_NAMESPACE
+from macaroonbakery.checkers.conditions import COND_DECLARED, STD_NAMESPACE
 from macaroonbakery.auth_context import ContextKey
```
Comment (rogpeppe) [inline macaroonbakery/checkers/declared.py]: I don't think that checkers should be importing from bakery (gives high likelihood of import cycles).
Maybe that's an argument for moving AuthContext into checkers.

File: macaroonbakery/oven.py [hunk 25]
```diff
@@ -60,7 +62,7 @@ def __init__(self, key=None, location=None, locator=None, ns=None,
             my_store = store.MemoryKeyStore()
             self.root_keystore_for_ops = lambda x: my_store
 
-    def macaroon(self, version, expiry, caveats, ops):
+    def macaroon(self, ctx, version, expiry, caveats, ops):
```
Comment (rogpeppe) [inline macaroonbakery/oven.py]: Please document argument and return types. Also, I'm not sure that we should be passing context around as freely as we do in Go - in pybakery, the context is tied more strongly to auth, but in Go it's used for network operations too.

In this case, I don't think that the context is being used for auth purposes, so should probably be omitted.
That probably applies to other Oven methods too.

File: macaroonbakery/store.py [hunk 26]
```diff
@@ -10,13 +10,13 @@ class MemoryOpsStore:
     def __init__(self):
         self._store = {}
 
-    def put_ops(self, key, ops, time):
+    def put_ops(self, ctx, key, ops, time):
```
Comment (rogpeppe) [inline macaroonbakery/store.py]: As mentioned in Oven, I'm not sure that the context is appropriate here.

File: macaroonbakery/tests/test_checkers.py [hunk 27]
```diff
@@ -8,19 +8,12 @@
 from pymacaroons import Macaroon, MACAROON_V2
 
 from macaroonbakery.auth_context import AuthContext
-from macaroonbakery.checkers.checkers import (
-    Checker, Namespace, CheckerInfo
+from macaroonbakery.checkers import (
```
Comment (rogpeppe) [inline macaroonbakery/tests/test_checkers.py]: import checkers ?

File: macaroonbakery/tests/test_checker.py [hunk 28]
```diff
@@ -0,0 +1,988 @@
...
+from collections import namedtuple
+import json
+from unittest import TestCase
+from datetime import datetime, timedelta
+import pytz
+
+from pymacaroons.verifier import Verifier, FirstPartyCaveatVerifierDelegate
+import pymacaroons
+
+from macaroonbakery import LATEST_BAKERY_VERSION
```
Comment (rogpeppe) [inline macaroonbakery/tests/test_checker.py]: import macaroonbakery ?

File: macaroonbakery/tests/test_checker.py [hunk 29]
```diff
@@ -0,0 +1,988 @@
...
+from macaroonbakery.bakery import ThirdPartyInfo
+from macaroonbakery.discharge import discharge
+from macaroonbakery.identity import (
+    IdentityClient, IdentityError, SimpleIdentity
+)
+from macaroonbakery.checkers import (
```
Comment (rogpeppe) [inline macaroonbakery/tests/test_checker.py]: import macaroonbakery.checkers ?

File: macaroonbakery/tests/test_checker.py [hunk 30]
```diff
@@ -0,0 +1,988 @@
...
+        if user is None:
+            raise IdentityError('no username declared')
+        return SimpleIdentity(user)
+
+_DISCHARGE_USER_KEY = ContextKey('user-key')
+
+_DischargeRecord = namedtuple('_DISCHARGE_RECORD', ['location', 'user'])
+
+
+class _Discharger(object):
```
Comment (rogpeppe) [inline macaroonbakery/tests/test_checker.py]: Maybe some doc comments on these helper types?

File: macaroonbakery/tests/test_macaroon.py [hunk 31]
```diff
@@ -5,27 +5,34 @@
 
 import nacl.utils
 
-from macaroonbakery import bakery, macaroon, codec
-from macaroonbakery.checkers.caveat import Caveat
+from macaroonbakery import LATEST_BAKERY_VERSION, BAKERY_V1, macaroon, codec
```
Comment (rogpeppe) [inline macaroonbakery/tests/test_macaroon.py]: import macaroonbakery ?

File: macaroonbakery/tests/test_macaroon.py [hunk 32]
```diff
@@ -5,27 +5,34 @@
...
+from macaroonbakery.bakery import ThirdPartyInfo
+from macaroonbakery.macaroon import ThirdPartyLocator
+from macaroonbakery.third_party import legacy_namespace, ThirdPartyCaveatInfo
+from macaroonbakery.checkers import Caveat
```
Comment (rogpeppe) [inline macaroonbakery/tests/test_macaroon.py]: import checkers ?

File: macaroonbakery/tests/test_namespace.py [hunk 33]
```diff
@@ -3,7 +3,7 @@
 
 from unittest import TestCase
 
-from macaroonbakery.checkers.namespace import Namespace, deserialize_namespace
+from macaroonbakery.checkers import Namespace, deserialize_namespace
```
Comment (rogpeppe) [inline macaroonbakery/tests/test_namespace.py]: import checkers ?

File: macaroonbakery/tests/test_oven.py [hunk 34]
```diff
@@ -6,13 +6,19 @@
 import copy
 from datetime import datetime, timedelta
 
-from macaroonbakery import bakery, checker, oven, store
+from macaroonbakery import (
```
Comment (rogpeppe) [inline macaroonbakery/tests/test_oven.py]: import macaroonbakery ?

File: macaroonbakery/third_party.py [hunk 35]
```diff
@@ -0,0 +1,53 @@
...
+
+def legacy_namespace():
+    ''' Standard namespace for pre-version3 macaroons.
+    '''
+    ns = checkers.Namespace(None)
+    ns.register(checkers.STD_NAMESPACE, '')
+    return ns
+
+
+class ThirdPartyCaveatInfo:
```
Comment (rogpeppe) [inline macaroonbakery/third_party.py]: Perhaps this should be a named tuple?

File: macaroonbakery/checker.py [hunk 36]
```diff
@@ -36,4 +49,388 @@ def __hash__(self):
...
+
+class Checker(object):
+    '''Checker wraps a Checker for first party caveat and adds authentication
+    and authorization checks.
+
+    It uses macaroons as authorization tokens but it is not itself responsible
+    for creating the macaroons
+    See the Oven type (TODO) for one way of doing that.
+    '''
+    def __init__(self, checker=None, authorizer=None, identity_client=None,
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/checker.py]: it will use the same instance each time then which might not be ideal
http://effbot.org/zone/default-values.htm

File: macaroonbakery/macaroon.py (hunk 22 above)
Comment (fabricematrat) [reply macaroonbakery/macaroon.py]: is there a mistake in the go implementation as the only place it is used is here https://github.com/go-macaroon-bakery/macaroon-bakery/blob/v2-unstable/bakery/macaroon.go#L303 where we just treansform it to string directly ? @rogpeppe

File: macaroonbakery/authorizer.py [hunk 37]
```diff
@@ -81,20 +81,31 @@ def __init__(self, get_acl, allow_public=False):
...
         allowed = [False] * len(ops)
         has_allow = isinstance(identity, ACLIdentity)
         for i, op in enumerate(ops):
             acl = self._get_acl(ctx, op)
             if has_allow:
                 allowed[i] = identity.allow(ctx, acl)
             else:
-                allowed[i] = self._allow_public and EVERYONE in acl
-        return allowed, None
+                if len(acl) == 0:
```
Comment (rogpeppe) [inline macaroonbakery/authorizer.py]: Why is this needed?

File: macaroonbakery/identity.py (hunk 14 above)
Comment (rogpeppe) [reply macaroonbakery/identity.py]: Return type still not formally documented.

File: macaroonbakery/authorizer.py [hunk 38]
```diff
@@ -88,13 +88,24 @@ def authorize(self, ctx, identity, ops):
...
         allowed = [False] * len(ops)
         has_allow = isinstance(identity, ACLIdentity)
         for i, op in enumerate(ops):
             acl = self._get_acl(ctx, op)
             if has_allow:
                 allowed[i] = identity.allow(ctx, acl)
             else:
-                allowed[i] = self._allow_public and EVERYONE in acl
-        return allowed, None
+                if acl is None:
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/authorizer.py]: Done.

File: macaroonbakery/checker.py [hunk 39]
```diff
@@ -2,6 +2,19 @@
 # Licensed under the LGPLv3, see LICENCE file for details.
 from collections import namedtuple
 
+import pyrfc3339
+
+from macaroonbakery.authorizer import ClosedAuthorizer
+from macaroonbakery.identity import NoIdentities
+from macaroonbakery.checkers import (
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/checker.py]: in a follow up pr.

File: macaroonbakery/checker.py [hunk 40]
```diff
@@ -36,4 +49,388 @@ def __hash__(self):
         return hash(self.__key())
 
 
-LOGIN_OP = Op(entity="login", action="login")
+LOGIN_OP = Op(entity='login', action='login')
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/checker.py]: done.

File: macaroonbakery/checker.py [hunk 41]
```diff
@@ -36,4 +49,388 @@ def __hash__(self):
...
+
+
+class Checker(object):
+    '''Checker wraps a Checker for first party caveat and adds authentication
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/checker.py]: done.

File: macaroonbakery/checker.py (hunk 36 above)
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/checker.py]: done.

File: macaroonbakery/checker.py [hunk 42]
```diff
@@ -36,4 +49,388 @@ def __hash__(self):
...
+    and authorization checks.
+
+    It uses macaroons as authorization tokens but it is not itself responsible
+    for creating the macaroons
+    See the Oven type (TODO) for one way of doing that.
+    '''
+    def __init__(self, checker=None, authorizer=None, identity_client=None,
+                 macaroon_opstore=None):
+        '''
+        :param checker: a first party check which has a
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/checker.py]: done.

File: macaroonbakery/checker.py [hunk 43]
```diff
@@ -36,4 +49,388 @@ def __hash__(self):
...
+        check_first_party_caveat method.
+        :param authorizer (Authorizer): used to check whether an authenticated
+        user is allowed to perform operations. If it is None, it will default
+        to a ClosedAuthorizer.
+        The identity parameter passed to authorizer.allow will always have been
+        obtained from a call to identity_client.declared_identity.
+        :param identity_client:	used for interactions with the external
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/checker.py]: done.

File: macaroonbakery/checker.py [hunk 44]
```diff
@@ -36,4 +49,388 @@ def __hash__(self):
...
+    of the AuthChecker is called, using the context passed in then.
+
+    To find out any declared identity without requiring a login,
+    use allow(ctx); to require authentication but no additional operations,
+    use allow(ctx, LOGIN_OP).
+    '''
+    def __init__(self, parent, macaroons):
+        '''
+
+        :param parent: used to check first party caveats.
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/checker.py]: done.

File: macaroonbakery/checker.py (hunk 1 above)
Comment (fabricematrat) [reply macaroonbakery/checker.py]: done.

File: macaroonbakery/checker.py [hunk 45]
```diff
@@ -36,4 +49,388 @@ def __hash__(self):
...
+        self._identity = None
+        self._identity_caveats = []
+        self.parent = parent
+        self._conditions = None
+
+    def _init(self, ctx):
+        if not self._executed:
+            self._init_once(ctx)
+            self._executed = True
+        return self._init_errors
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/checker.py]: done.

File: macaroonbakery/checker.py (hunk 2 above)
Comment (fabricematrat) [reply macaroonbakery/checker.py]: done.

File: macaroonbakery/checker.py (hunk 3 above)
Comment (fabricematrat) [reply macaroonbakery/checker.py]: done.

File: macaroonbakery/checker.py (hunk 4 above)
Comment (fabricematrat) [reply macaroonbakery/checker.py]: done.

File: macaroonbakery/codec.py (hunk 5 above)
Comment (fabricematrat) [reply macaroonbakery/codec.py]: in a follow up pr

File: macaroonbakery/discharge.py (hunk 6 above)
Comment (fabricematrat) [reply macaroonbakery/discharge.py]: in a follow up pr

File: macaroonbakery/discharge.py (hunk 7 above)
Comment (fabricematrat) [reply macaroonbakery/discharge.py]: done.

File: macaroonbakery/discharge.py (hunk 8 above)
Comment (fabricematrat) [reply macaroonbakery/discharge.py]: done.

File: macaroonbakery/discharge.py (hunk 9 above)
Comment (fabricematrat) [reply macaroonbakery/discharge.py]: done.

File: macaroonbakery/discharge.py (hunk 10 above)
Comment (fabricematrat) [reply macaroonbakery/discharge.py]: done.

File: macaroonbakery/error.py (hunk 11 above)
Comment (fabricematrat) [reply macaroonbakery/error.py]: done.

File: macaroonbakery/error.py (hunk 12 above)
Comment (fabricematrat) [reply macaroonbakery/error.py]: done.

File: macaroonbakery/identity.py (hunk 13 above)
Comment (fabricematrat) [reply macaroonbakery/identity.py]: done.

File: macaroonbakery/identity.py (hunk 14 above)
Comment (fabricematrat) [reply macaroonbakery/identity.py]: done.

File: macaroonbakery/identity.py (hunk 15 above)
Comment (fabricematrat) [reply macaroonbakery/identity.py]: done.

File: macaroonbakery/identity.py (hunk 16 above)
Comment (fabricematrat) [reply macaroonbakery/identity.py]: done.

File: macaroonbakery/keys.py (hunk 17 above)
Comment (fabricematrat) [reply macaroonbakery/keys.py]: done.

File: macaroonbakery/macaroon.py (hunk 18 above)
Comment (fabricematrat) [reply macaroonbakery/macaroon.py]: done.

File: macaroonbakery/macaroon.py (hunk 21 above)
Comment (fabricematrat) [reply macaroonbakery/macaroon.py]: done.

File: macaroonbakery/macaroon.py (hunk 19 above)
Comment (fabricematrat) [reply macaroonbakery/macaroon.py]: done.

File: macaroonbakery/macaroon.py (hunk 20 above)
Comment (fabricematrat) [reply macaroonbakery/macaroon.py]: left over.removed.

File: macaroonbakery/macaroon.py (hunk 23 above)
Comment (fabricematrat) [reply macaroonbakery/macaroon.py]: fixed the whole algo in the next PR coming.

File: macaroonbakery/macaroon.py (hunk 22 above)
Comment (fabricematrat) [reply macaroonbakery/macaroon.py]: fixed

File: macaroonbakery/checkers/declared.py (hunk 24 above)
Comment (fabricematrat) [reply macaroonbakery/checkers/declared.py]: yep

File: macaroonbakery/store.py (hunk 26 above)
Comment (fabricematrat) [reply macaroonbakery/store.py]: done.

File: macaroonbakery/checker.py [hunk 46]
```diff
@@ -36,4 +47,391 @@ def __hash__(self):
...
+    '''Authorizes operations with respect to a user's request.
+
+    The identity is authenticated only once, the first time any method
+    of the AuthChecker is called, using the context passed in then.
+
+    To find out any declared identity without requiring a login,
+    use allow(ctx); to require authentication but no additional operations,
+    use allow(ctx, LOGIN_OP).
+    '''
+    mutex = Lock()
```
Comment (rogpeppe) [inline macaroonbakery/checker.py]: This doesn't seem quite right - this allocates a global lock, but we only need a lock on a given AuthChecker instance.

File: macaroonbakery/checker.py [hunk 47]
```diff
@@ -36,4 +47,391 @@ def __hash__(self):
...
+        self._macaroons = macaroons
+        self._init_errors = []
+        self._executed = False
+        self._identity = None
+        self._identity_caveats = []
+        self.parent = parent
+        self._conditions = None
+
+    def _init(self, ctx):
+        try:
```
Comment (rogpeppe) [inline macaroonbakery/checker.py]: You don't need to call acquire and release manually:

    with self.mutex:
        if not etc

should work fine (assuming the lock is in the instance not the class).

File: macaroonbakery/third_party.py [hunk 48]
```diff
@@ -0,0 +1,55 @@
...
+        the caveat id.
+        @params Namespace object that holds the namespace of the first party
+        that created the macaroon, as encoded by the party that added the
+        third party caveat.
+        '''
+        return super(ThirdPartyCaveatInfo, cls).__new__(
+            cls, condition, first_party_public_key, third_party_key_pair,
+            root_key, caveat, version, ns)
+
+    def __eq__(self, other):
```
Comment (rogpeppe) [inline macaroonbakery/third_party.py]: Why is this needed? ISTM that equality works fine without it.

File: macaroonbakery/third_party.py [hunk 49]
```diff
@@ -0,0 +1,55 @@
...
+        @param caveat holds the full encoded base64 string caveat id from
+        which all the other fields are derived.
+        @param version holds the version that was used to encode
+        the caveat id.
+        @params Namespace object that holds the namespace of the first party
+        that created the macaroon, as encoded by the party that added the
+        third party caveat.
+        '''
+        return super(ThirdPartyCaveatInfo, cls).__new__(
+            cls, condition, first_party_public_key, third_party_key_pair,
```
Comment (rogpeppe) [inline macaroonbakery/third_party.py]: Maybe use named arguments here so we don't run the risk of muddling them up if they're declared in a different order in the namedtuple?

File: macaroonbakery/third_party.py (hunk 48 above)
Comment (fabricematrat) [reply macaroonbakery/third_party.py]: nope you can try to comment it and run the test, you will see failing tests.
